File: testing/unit-tests/ruby/RubyPokemonTest/Gen1PokemonTest.rb
#!/usr/bin/ruby
#
# Copyright (c) 2017 Nicholas Corgan (n.corgan@gmail.com)
#
# Distributed under the MIT License (MIT) (See accompanying file LICENSE.txt
# or copy at http://opensource.org/licenses/MIT)
#

require_relative "PokemonTest"
require_relative "PokemonTestParams"

require "PKMN"

require "minitest/autorun"
require "os"
require "pathname"

class Gen1PokemonTest < PokemonTest
    def _gen1_pokemon_test(game_params)
        pokemon = PKMN::Pokemon.new(game_params[1], game_params[0], "", 30)

        test_params = PokemonTestParams.new
        test_params.valid_ball = "Great Ball"
        test_params.invalid_balls = ["Great Ball"]
        test_params.valid_item = "Potion"
        test_params.invalid_items = ["Potion"]
        test_params.expected_original_location = "Special"
        test_params.valid_locations = ["Route 1"]
        test_params.invalid_locations = ["Route 1"]
        test_params.valid_moves = ["Slash", "Flamethrower", "Tail Whip", "Fire Blast"]
        test_params.invalid_moves = ["Return"]
        test_params.valid_original_games = ["Red"]
        test_params.invalid_original_games = ["Red"]

        pokemon_test_common(pokemon, test_params)
    end

    [["Red", "Charmander"],
     ["Blue", "Squirtle"],
     ["Yellow", "Bulbasaur"]].each do |game_params|
        define_method("test_gen1_pokemon_#{game_params[0]}") do
            _gen1_pokemon_test(game_params)
        end
    end
end

File: testing/unit-tests/c/c_pokemon_test/gen1_pokemon_test.c
/*
 * Copyright (c) 2017-2018 Nicholas Corgan (n.corgan@gmail.com)
 *
 * Distributed under the MIT License (MIT) (See accompanying file LICENSE.txt
 * or copy at http://opensource.org/licenses/MIT)
 */

#include "c_test_common.h"

#include "pokemon_test.h"

#include <pkmn-c.h>

static void gen1_pokemon_test(
    enum pkmn_species species,
    enum pkmn_game game
)
{
    enum pkmn_error error = PKMN_ERROR_NONE;
    struct pkmn_pokemon pokemon =
    {
        .species = PKMN_SPECIES_NONE,
        .game = PKMN_GAME_NONE,
        .p_internal = NULL
    };

    error = pkmn_pokemon_init(
                species,
                game,
                "",
                30,
                &pokemon
            );
    PKMN_TEST_ASSERT_SUCCESS(error);
    TEST_ASSERT_NOT_NULL(pokemon.p_internal);

    pkmn_test_values_t test_values =
    {
        .valid_ball = PKMN_BALL_GREAT_BALL,
        .p_invalid_balls = (enum pkmn_ball[]){PKMN_BALL_GREAT_BALL, PKMN_BALL_NONE},

        .valid_item = PKMN_ITEM_POTION,
        .p_invalid_items = (enum pkmn_item[]){PKMN_ITEM_POTION, PKMN_ITEM_NONE},

        .expected_original_location = "Special",
        .valid_locations = (char*[]){"Route 1", NULL},
        .invalid_locations = (char*[]){"Route 1", NULL},

        .moves = (char*[]){"Slash", "Flamethrower", "Tail Whip", "Fire Blast", NULL},
        .invalid_moves = (char*[]){"Return", NULL},

        .valid_original_games = (enum pkmn_game[]){PKMN_GAME_RED, PKMN_GAME_NONE},
        .invalid_original_games = (enum pkmn_game[]){PKMN_GAME_RED, PKMN_GAME_NONE}
    };

    pokemon_test_common(
        &pokemon,
        &test_values
    );

    // Make sure we can read the relevant attributes.
    int catch_rate = 0;
    error = pkmn_pokemon_get_numeric_attribute(
                &pokemon,
                "Catch rate",
                &catch_rate
            );
    PKMN_TEST_ASSERT_SUCCESS(error);

    error = pkmn_pokemon_free(&pokemon);
    PKMN_TEST_ASSERT_SUCCESS(error);
}

void red_pokemon_test()
{
    gen1_pokemon_test(PKMN_SPECIES_CHARMANDER, PKMN_GAME_RED);
}

void blue_pokemon_test()
{
    gen1_pokemon_test(PKMN_SPECIES_SQUIRTLE, PKMN_GAME_BLUE);
}

void yellow_pokemon_test()
{
    gen1_pokemon_test(PKMN_SPECIES_BULBASAUR, PKMN_GAME_YELLOW);
}
